e2e_playwright/conftest.py

```diff
@@ -248,6 +248,32 @@ def app(page: Page, app_port: int) -> Page:
     return page
 
 
+@pytest.fixture(scope="function")
+def app_with_params(page: Page, app_port: int, request) -> Tuple[Page, Dict]:
+    """Fixture that opens the app with additional query parameters.
+    The query parameters are passed as a dictionary in the 'query_params' key of the request.
+    """
+    query_params = request.param
+
+    from urllib import parse
+
+    query_parts = []
+    for key, value in query_params.items():
+        if isinstance(value, (list, tuple)):
+            for item in value:
+                query_parts.append(f"{key}={parse.quote(str(item))}")
+        else:
+            query_parts.append(f"{key}={parse.quote(str(value))}")
+
+    query_string = "&".join(query_parts)
+
+    url = f"http://localhost:{app_port}/?{query_string}"
```

Comment:
This seems like it's being done really manually. I'd guess there are library implementations that should allow you to construct URLs with a given dictionary of query params without so much effort